presto-main/src/main/java/io/prestosql/operator/scalar/ArrayAnyMatchFunction.java

```diff
@@ -0,0 +1,219 @@
+            @SqlType("function(T, boolean)") AnyMatchLongLambda function)
+    {
+        boolean hasNullResult = false;
+        int positionCount = arrayBlock.getPositionCount();
+        for (int i = 0; i < positionCount; i++) {
+            Long element = null;
+            if (!arrayBlock.isNull(i)) {
+                element = elementType.getLong(arrayBlock, i);
+            }
+            Boolean match = function.apply(element);
+            if (TRUE.equals(match)) {
+                return true;
+            }
+            if (match == null) {
+                hasNullResult = true;
+            }
+        }
+        if (hasNullResult) {
+            return null;
+        }
+        return false;
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = double.class)
+    @SqlType(StandardTypes.BOOLEAN)
+    @SqlNullable
+    public static Boolean anyMatchDouble(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block arrayBlock,
+            @SqlType("function(T, boolean)") AnyMatchDoubleLambda function)
+    {
+        boolean hasNullResult = false;
+        int positionCount = arrayBlock.getPositionCount();
+        for (int i = 0; i < positionCount; i++) {
+            Double element = null;
+            if (!arrayBlock.isNull(i)) {
+                element = elementType.getDouble(arrayBlock, i);
+            }
+            Boolean match = function.apply(element);
+            if (TRUE.equals(match)) {
+                return true;
+            }
+            if (match == null) {
+                hasNullResult = true;
+            }
+        }
+        if (hasNullResult) {
+            return null;
+        }
+        return false;
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = boolean.class)
+    @SqlType(StandardTypes.BOOLEAN)
+    @SqlNullable
+    public static Boolean anyMatchBoolean(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block arrayBlock,
+            @SqlType("function(T, boolean)") AnyMatchBooleanLambda function)
+    {
+        boolean hasNullResult = false;
+        int positionCount = arrayBlock.getPositionCount();
+        for (int i = 0; i < positionCount; i++) {
+            Boolean element = null;
+            if (!arrayBlock.isNull(i)) {
+                element = elementType.getBoolean(arrayBlock, i);
+            }
+            Boolean match = function.apply(element);
+            if (TRUE.equals(match)) {
+                return true;
+            }
+            if (match == null) {
+                hasNullResult = true;
+            }
+        }
+        if (hasNullResult) {
+            return null;
+        }
+        return false;
+    }
+
+    @FunctionalInterface
+    public interface AnyMatchBlockLambda
+            extends LambdaFunctionInterface
+    {
+        Boolean apply(Block x);
+    }
+
+    @FunctionalInterface
```

Comment:
Instead of duplicating these interfaces across the multiple functions, pull them out and reuse them. There's nothing that ties them to each of the functions -- all that matters is that they are functional interfaces for XXX -> YYY, for some XXX and YYY types.

Also, I'd rename them to `SliceToBooleanFunction`, `LongToBooleanFunction`, `DoubleToBooleanFunction` and `BooleanToBooleanFunction` (the "lambda" is the syntactic construct of what gets passed in, but the methods that use them take a "function")

Reply:
@martint 
I'd also pull out interfaces in `ArrayFilterFunction` because they are same as those in `ArrayAllMatchFunction` and `ArrayAnyMatchFunction`.

Two questions:
1. Should I put these interfaces in package `io.prestosql.operator.scalar`?
2. Should I separate the changes in two commits? One for extracting and renaming interfaces from `ArrayFilterFunction`, the other one for adding match functions.

Reply:
Yes, both seem reasonable